File: 2_add_two_numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        result = marker = ListNode()
        remainder = 0

        while l1 and l2:
            x = l1.val
            y = l2.val

            summ = x + y + remainder
            if summ < 10:
                result.next = ListNode(summ)
                result = result.next
                remainder = 0
            else:
                deno = summ % 10
                remainder = summ // 10
                result.next = ListNode(deno)
                result = result.next

            l1 = l1.next
            l2 = l2.next

        if l1:
            while l1:
                summ = l1.val + remainder
                if summ < 10:
                    result.next = ListNode(summ)
                    result = result.next
                    remainder = 0
                else:
                    deno = summ % 10
                    remainder = summ // 10
                    result.next = ListNode(deno)
                    result = result.next
                l1 = l1.next

        if l2:
            while l2:
                summ = l2.val + remainder
                if summ < 10:
                    result.next = ListNode(summ)
                    result = result.next
                    remainder = 0
                else:
                    deno = summ % 10
                    remainder = summ // 10
                    result.next = ListNode(deno)
                    result = result.next
                l2 = l2.next
        if remainder > 0:
            result.next = ListNode(remainder)
            result = result.next
            result.next = None


        return marker.next

File: linked_list/2_add_two_numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode res = new ListNode(-1);
        ListNode cur = res;
        int x = 0, y = 0, sum = 0, carry = 0;

        while (l1 != null || l2 != null || carry != 0) {
            x = (l1 != null) ? l1.val : 0;
            y = (l2 != null) ? l2.val : 0;
            sum = x + y + carry;
            carry = sum / 10;

            cur.next = new ListNode(sum % 10);
            cur = cur.next;

            if (l1 != null) l1 = l1.next;
            if (l2 != null) l2 = l2.next;
        }

        return res.next;
    }
}